tests/test_search.py

```diff
@@ -67,7 +68,16 @@ def load_searchindex(path: Path) -> Any:
     assert searchindex.startswith('Search.setIndex(')
     assert searchindex.endswith(')')
 
-    return json.loads(searchindex[16:-1])
+    indices = ast.literal_eval(
+        searchindex[16:-1]
+        .replace('new Map', '')
+        .replace('null', 'None')
+        .replace('false', 'False')
+    )
```

Comment:
This is fragile and inelegant; it's a compromise to avoid introducing the complexity of a Python-implemented JavaScript parser dependency given the use of `new Map(...)` within `searchindex.js` files.

Reply:
Another alternative would be to dump the index as JSON, then reload it into the actual format we want. Would almost certainly be a little slower, but probably not disastrously so.

Reply:
Could you explain that in a little more detail?  I think I understand what you mean, but I'd like to make sure.

Reply:
Yeah I mean actually just writing out a JSON file and having the frontend code load it seperately via a fetch call. I'm not actually sure it would be slower, now that I think about it-- this post suggests this will be faster on the interpreter side: https://v8.dev/blog/cost-of-javascript-2019#json

Reply:
Ok, thanks! (I wasn't entirely sure whether you were suggesting JSON format output only for the unit test code - using it in both seems more robust).

Let's give `searchindex.json` a try, in that case.  It'll make a few things simpler, although the `searchtools.js` code may lose some of the readability improvements (we'll want to begin/reintroduce using `obj.hasOwnProperty` again in a few places).

We may also want to be careful about where we place the `fetch` call so that the browser doesn't block for unnecessarily long before invoking that (perhaps especially relevant for the case of downstream themes that include their own scripting that may also run on each page load).

Reply:
> Let's give searchindex.json a try, in that case. It'll make a few things simpler, although the searchtools.js code may lose some of the readability improvements (we'll want to begin/reintroduce using obj.hasOwnProperty again in a few places).

Personally I would read the search index as JSON, then convert it into a Map (i.e. building on the work you already did).

> We may also want to be careful about where we place the fetch call so that the browser doesn't block for unnecessarily long before invoking that (perhaps especially relevant for the case of downstream themes that include their own scripting that may also run on each page load).

Yeah that part might be fiddly. Some types of interactivity might not be fully supported until this operation is complete, so this might complicate things. If it gets too out of control, maybe best to go back to the previous approach.

Reply:
> > Let's give searchindex.json a try, in that case. It'll make a few things simpler, although the searchtools.js code may lose some of the readability improvements (we'll want to begin/reintroduce using obj.hasOwnProperty again in a few places).
> 
> Personally I would read the search index as JSON, then convert it into a Map (i.e. building on the work you already did).

I'd prefer to avoid multiple read passes over the search index data on the client-side, if possible.  Similarly that makes me reluctant to use the `reviver` function in `JSON.parse`, and I considered it a downside of the freeze-and-deproto code I had implemented in #13098.

Aside from code clarity, do you anticipate many benefits from the transformation to `Map` datastructures?

Reply:
> Aside from code clarity, do you anticipate many benefits from the transformation to Map datastructures?

Not really, my guess is that getting entries out of a map will be a little faster (since it's a datastructure purely oriented around storing and retrieving keys/values) but probably not noticeably so. I think it's at least worth exploring though. Probably just a couple extra lines of code after you've loaded the JSON structure.

Reply:
Generally I feel inclined to disentangle the two aspects of this pull request: adding support for `__proto__` as an indexed/queryable term, and preventing `__proto__` queries from accessing JavaScript object prototypes.

The latter seems more important (it's a bug) - the former is a feature request that hasn't really been requested by anyone other than myself during development.

Reply:
I've extracted the bugfix part of this changeset into #13153; I don't have any plans currently to add feature support for term/title-term matching of the query term `__proto__` itself.